src/instruction.rs

```diff
@@ -1080,19 +1131,91 @@ mod test_instruction_display {
 #[derive(Clone, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
 pub enum Qubit {
     Fixed(u64),
+    #[serde(skip)] // todo
+    Placeholder(QubitPlaceholder),
     Variable(String),
 }
 
+impl Qubit {
+    fn resolve_placeholder<R>(&mut self, resolver: R)
+    where
+        R: Fn(&QubitPlaceholder) -> Option<u64>,
+    {
+        if let Qubit::Placeholder(placeholder) = self {
+            if let Some(resolved) = resolver(placeholder) {
+                *self = Qubit::Fixed(resolved);
+            }
+        }
+    }
+
+    // For comparison, this takes a more active approach than the above, in that it can allocate new indices
+    // using the placeholder. This seems unnecessary and so I've commented it out, but leaving it here
+    // for commentary while in draft
+    // fn resolve_placeholder<I, R>(
+    //     &mut self,
+    //     indices: &mut I,
+    //     resolver: R,
+    // ) -> Result<Option<(QubitPlaceholder, u64)>, ()>
+    // where
+    //     I: Iterator<Item = u64>,
+    //     R: Fn(QubitPlaceholder) -> Option<u64>,
+    // {
+    //     if let Qubit::Placeholder(placeholder) = self {
+    //         if let Some(resolved) = resolver(placeholder.clone()) {
+    //             *self = Qubit::Fixed(resolved);
+    //             Ok(None)
+    //         } else if let Some(next) = indices.next() {
+    //             let return_value = Some((placeholder.clone(), next));
+    //             *self = Qubit::Fixed(next);
+    //             Ok(return_value)
+    //         } else {
+    //             Err(())
+    //         }
+    //     } else {
+    //         Ok(None)
+    //     }
+    // }
+}
+
 impl fmt::Display for Qubit {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         use Qubit::*;
         match self {
             Fixed(value) => write!(f, "{}", value),
+            Placeholder(_) => todo!(),
             Variable(value) => write!(f, "{}", value),
         }
     }
 }
 
+type QubitPlaceholderInner = Arc<Pin<Box<usize>>>;
```

Comment:
thankfully, got rid of it. A fixed memory address isn't important if the user never attempts to compare addresses themselves; since they can't access the inner types of the placeholders, that shouldn't be a problem.